Truncate metadata

riiwo opened the pull request:
When metadata is too big, we want to truncate it so the log parsing
keeps working.

Going through each metadata key-value pair and checking its size.
For big lists and maps, some of the items will be discarded. And
long strings will be truncated.

urmastalimaa commented:
Parsing what has been truncated in the logs might be quite tricky. For FE logs we have the "-pruned-" marker (also see recent https://github.com/salemove/client-logger-js/pull/16), here we either want a similar system, or, as there is no front-end -> back-end data transfer we can be a bit more lenient.

More precisely, could we split
`{"message", big_metadata}`
into
* `{"message (pruned)", metadata_1}`
* `{"message (continued)", metadata_2}`
* ....
* `{"message (continued)", metadata_n}`

How difficult would that be?
As a simpler alternative, we could add (pruned) to the message and drop all non-essential metadata (application, kube_pod, etc) and let the service maintainer fix the problem.

riiwo commented:
> Parsing what has been truncated in the logs might be quite tricky. For FE logs we have the "-pruned-" marker (also see recent [salemove/client-logger-js#16](https://github.com/salemove/client-logger-js/pull/16)), here we either want a similar system, or, as there is no front-end -> back-end data transfer we can be a bit more lenient.
> 
> More precisely, could we split
> `{"message", big_metadata}`
> into
> 
> * `{"message (pruned)", metadata_1}`
> * `{"message (continued)", metadata_2}`
> * ....
> * `{"message (continued)", metadata_n}`
> 
> How difficult would that be?
> As a simpler alternative, we could add (pruned) to the message and drop all non-essential metadata (application, kube_pod, etc) and let the service maintainer fix the problem.

Added pruned.

With the current solution, none of the metadata top-level keys are dropped. Dropping start inside nesting, so this should be a none issue, that stuff gets lost. 

Not sure how to make the other solution work easily.

urmastalimaa commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +75,74 @@ defmodule LogstashLoggerFormatter do
+      item
+      |> Enum.take(keep_items_count)
+      |> Enum.map(&maybe_truncate_item/1)
+    if keep_items_count < list_length, do: truncated_list ++ ["-pruned-"], else: truncated_list
```
This will cause log parsing errors if there are objects inside the list, see https://github.com/salemove/client-logger-js/pull/16

urmastalimaa commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +75,74 @@ defmodule LogstashLoggerFormatter do
+  defp truncate_metadata(md) when is_struct(md), do: truncate_metadata(Map.from_struct(md))
+
+  defp truncate_metadata(md) when is_map(md) do
+    Enum.reduce(md, %{}, fn {key, value}, acc ->
```
As most of the logs are hopefully not past the max metadata size, do you think an early exit here if the whole object is less than max size would be useful?

indrekj commented:
Can you explain what part of the system is failing with big logs? Is it fluentd, elasticsearch, something else? Is it configurable and what are the values/maximums rights? Is it one metadata field, all together, the message altogether?

I understand the idea here but I have no idea how expensive actually `Kernel.byte_size(Kernel.inspect(md))` is if we do that for every log message. It could have a significant impact on a service performance that generates a lot of logs.

riiwo commented:
> Can you explain what part of the system is failing with big logs? Is it fluentd, elasticsearch, something else? Is it configurable and what are the values/maximums rights? Is it one metadata field, all together, the message altogether?
> 
> I understand the idea here but I have no idea how expensive actually `Kernel.byte_size(Kernel.inspect(md))` is if we do that for every log message. It could have a significant impact on a service performance that generates a lot of logs.

I don't know for sure, but I assume elasticsearch fails to parse too big logs. Original ticket can be found here: https://salemove.atlassian.net/browse/EN-5825 . And currently engagement-router is the one that has too big logs.
I'd assume it is usually one or two metadata fields. Message fields themselves are fine since elixir logger has truncate built-in for that (default 8096 bytes).

riiwo commented:
Ran some benchmarks on long string (~31kB):

```
Name                         ips        average  deviation         median         99th %
Kernel.inspect            737.24        1.36 ms    ±18.19%        1.36 ms        2.18 ms
+ Kernel.byte_size        731.00        1.37 ms    ±19.89%        1.36 ms        2.26 ms

Comparison: 
Kernel.inspect            737.24
+ Kernel.byte_size        731.00 - 1.01x slower +0.0116 ms
```
So byte_size is very fast but inspect can take more than 1 ms when very large data.

Same data converted to array, made inspect 8x slower.
```
Name                         ips        average  deviation         median         99th %
+ Kernel.byte_size        112.41        8.90 ms    ±10.06%        8.54 ms       11.94 ms
Kernel.inspect            110.89        9.02 ms    ±14.99%        8.54 ms       16.01 ms

Comparison: 
+ Kernel.byte_size        112.41
Kernel.inspect            110.89 - 1.01x slower +0.122 ms
```

Also map was about the same speed:
```
Name                         ips        average  deviation         median         99th %
Kernel.inspect            139.26        7.18 ms    ±10.55%        6.88 ms        9.84 ms
+ Kernel.byte_size        139.13        7.19 ms    ±10.19%        6.92 ms        9.63 ms

Comparison: 
Kernel.inspect            139.26
+ Kernel.byte_size        139.13 - 1.00x slower +0.00690 ms
```

Small maps and arrays took only microseconds, so not expecting any issues there:
```
Name                         ips        average  deviation         median         99th %
Kernel.inspect            2.71 K      369.28 μs    ±22.15%         329 μs      688.25 μs
+ Kernel.byte_size        2.67 K      375.09 μs    ±22.48%         332 μs      697.34 μs

Comparison: 
Kernel.inspect            2.71 K
+ Kernel.byte_size        2.67 K - 1.02x slower +5.81 μs
```

indrekj commented:
> I don't know for sure, but I assume elasticsearch fails to parse too big logs.

Could you find out the limitations? Currently, it seems we're just truncating using arbitrary rules and just hoping that it was enough.

> but inspect can take more than 1 ms when very large data

Can you also see how `Jason.encode!` compares to inspect?

riiwo commented:
Comparison:
```
Name                                ips        average  deviation         median         99th %
+ Kernel.byte_size map           2.18 K      458.54 μs    ±16.84%         425 μs         763 μs
Jason.encode! map                2.04 K      489.38 μs    ±19.69%         445 μs      840.16 μs
+ Kernel.byte_size string        1.79 K      557.71 μs    ±16.33%         515 μs      908.55 μs
+ Kernel.byte_size list          1.75 K      571.29 μs    ±17.95%         519 μs         969 μs
Jason.encode! string             1.72 K      581.15 μs    ±24.54%         516 μs      986.54 μs
Jason.encode! list               1.69 K      593.40 μs    ±24.23%         523 μs     1213.76 μs

Comparison: 
+ Kernel.byte_size map           2.18 K
Jason.encode! map                2.04 K - 1.07x slower +30.84 μs
+ Kernel.byte_size string        1.79 K - 1.22x slower +99.16 μs
+ Kernel.byte_size list          1.75 K - 1.25x slower +112.75 μs
Jason.encode! string             1.72 K - 1.27x slower +122.61 μs
Jason.encode! list               1.69 K - 1.29x slower +134.85 μs
```

riiwo commented:
According to this: https://www.elastic.co/guide/en/app-search/current/limits.html
Maximum Document Size: 100KB [configurable in 7.7+] so 10k is not arbitrary but 4k is hopeful that there aren't more than 2 metadata points that need truncating. I could add add some calculation but I don't think it will be that useful and it should be quire rare that it would still faili.

indrekj commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +76,73 @@ defmodule LogstashLoggerFormatter do
+    max(floor((@max_metadata_item_size / byte_size) * length), 1)
+  end
+
+  defp metadata_item_too_big(md), do: metadata_byte_size(md) > @max_metadata_item_size
```
you can also use question marks for boolean functions `metadata_item_too_big?(md),`

indrekj commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +76,73 @@ defmodule LogstashLoggerFormatter do
+  end
+  defp truncate_metadata(md), do: md
+
+  defp maybe_truncate_item(md) when is_number(md) when is_atom(md), do: md
```
`when is_number(md) or is_atom(md),`

indrekj commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +76,73 @@ defmodule LogstashLoggerFormatter do
+
+  defp maybe_truncate_item(md) when is_number(md) when is_atom(md), do: md
+  defp maybe_truncate_item(md) when is_struct(md) do
+    if metadata_item_too_big?(md), do: truncate_item(Map.from_struct(md)), else: md
```
This `from_struct` and maybe above as well change the logging behaviour. See the comment for function `  defp format_metadata(%_{} = md) do`. Basically if it's a struct then so far we've let Jason encoder to take care of formatting it.  E.g. we do that to convert TapperId: https://github.com/salemove/engagement-router/blob/master/lib/tracing/encoder.ex#L1 . This change may break this.

riiwo replied:
What could be a solution here?

Maybe skip everything in case of a struct?

TBH I didn't think this is ever used, I added it just in case.

indrekj replied:
could start with skipping and see if that's fine for now

indrekj commented:
Generally, I don't really like truncating backend service logs because it can have an impact on performance (2x encoding, traversing through the metadata, etc). I think it's good that there's an early return in case the payload is not too large, which allows us to skip going through the metadata in most of the cases (still 2x encoding though). I added a couple of comments. No other immediate concerns.

indrekj commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +76,73 @@ defmodule LogstashLoggerFormatter do
+  end
+
+  defp truncate_item(item) do
+    String.slice(Jason.encode!(item), 0..@max_metadata_item_size - 1) <> " (-pruned-)"
```
Should use `@encode_fn` here to support both Poison and Jason (see how it's used above).

urmastalimaa reviewed (approved):
Yeah, good work on getting this to work, but the whole thing is a bit unfortunate.

indrekj commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +76,70 @@ defmodule LogstashLoggerFormatter do
   defp metadata_key(:application), do: :otp_application
   defp metadata_key(key), do: key
 
+  defp truncate_metadata(md) when is_struct(md) do
```
this one also has from_struct call

indrekj commented on lib/logstash_logger_formatter.ex:
```diff
@@ -73,6 +76,70 @@ defmodule LogstashLoggerFormatter do
+  defp truncate_metadata(md) when is_struct(md) do
+    if metadata_too_big?(md), do: truncate_metadata(Map.from_struct(md)), else: md
+  end
+  defp truncate_metadata(md) when is_map(md) do
```
should this also include `is_list` or does the `format_metadata` already convert keyword lists into maps?

riiwo replied:
Already a map, so this and is_struct parts should never happen

riiwo commented:
!check

riiwo commented:
!check
